packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +51,407 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
-    const mergedSchema = {
-      ...removeUndefinedValues(propertySchemaFromType),
-      ...removeUndefinedValues(rest as Omit<ObjectSchemaProperty, "name">),
-    } as ObjectSchemaProperty & { type: PropertyTypeName };
-    if (mergedSchema.type === "mixed" || mergedSchema.objectType === "mixed") {
-      assert(mergedSchema.optional !== false, "Mixed values should be declared as optional");
+  const normalizedSchema =
+    typeof schema === "string"
+      ? normalizePropertySchemaString(name, schema)
+      : normalizePropertySchemaObject(name, schema);
+
+  if (isPrimaryKey) {
+    assert(!normalizedSchema.optional, errMessage(name, "Optional properties cannot be used as a primary key."));
+    normalizedSchema.indexed = true;
+  }
+
+  return normalizedSchema;
+}
+
+/**
+ * Transform and validate a user-provided property schema that is using
+ * the shorthand string notation into its canonical form.
+ */
+function normalizePropertySchemaString(name: ObjectAndPropertyName, schema: string): CanonicalObjectSchemaProperty {
+  assert(schema.length > 0, errMessage(name, "The type must be specified."));
+
+  let type = "";
+  let objectType: string | undefined;
+  let optional = false;
+
+  if (endsWithCollection(schema)) {
+    const end = schema.substring(schema.length - 2);
+    type = COLLECTION_SHORTHAND_TO_NAME[end];
+
+    schema = schema.substring(0, schema.length - 2);
+    assert(schema.length > 0, errMessage(name, `The element type must be specified. (Example: 'int${end}')`));
+
+    const isNestedCollection = endsWithCollection(schema);
+    assert(!isNestedCollection, errMessage(name, "Nested collections are not supported."));
+  }
+
+  if (schema.endsWith("?")) {
+    optional = true;
+
+    schema = schema.substring(0, schema.length - 1);
+    assert(schema.length > 0, errMessage(name, "The type must be specified. (Examples: 'int?', 'int?[]')"));
+
+    const usingOptionalOnCollection = endsWithCollection(schema);
+    assert(
+      !usingOptionalOnCollection,
+      errMessage(
+        name,
+        "Collections cannot be optional. To allow elements of the collection to be optional, use '?' after the element type. (Examples: 'int?[]', 'int?{}', 'int?<>')",
+      ),
+    );
+  }
+
+  if (isPrimitive(schema)) {
+    if (isCollection(type)) {
+      objectType = schema;
+    } else {
+      type = schema as PropertyTypeName;
     }
-    const result: CanonicalObjectSchemaProperty = {
-      indexed: false,
-      optional: false,
-      mapTo: name,
-      ...mergedSchema,
+  } else if (isCollection(schema)) {
+    error(name, "Cannot use the collection name. (Examples: 'int[]' (list), 'int{}' (dictionary), 'int<>' (set))");
+  } else if (schema === "object") {
+    error(name, "To define a relationship, use either 'ObjectName' or { type: 'object', objectType: 'ObjectName' }");
+  } else if (schema === "linkingObjects") {
```

Comment:
Currently we don't error or warn if a user types "linkingObject" (singular). If banning it, it would potentially be a breaking change if somebody has used that as their user-defined type, but what are your thoughts on dealing with that potential typo?

Reply:
Do you mean you think we should require users to capitalize the object names?

Reply:
Yes, I think so.  Maybe we can bring that up in a round table discussion

Reply:
Yeah let's do that